lib/staging/tls/openssl_util.cpp

```diff
@@ -0,0 +1,518 @@
+            if (signature_len == 0) {
+                log_error("i2d_ECDSA_SIG");
+            }
+        } else {
+            log_error("ECDSA_SIG_set0");
+        }
+    }
+
+    BN_free(rbn);
+    BN_free(sbn);
+    ECDSA_SIG_free(signature);
+    return {DER_Signature_ptr(sig_p, &DER_Signature_free), signature_len};
+};
+
+bool signature_to_bn(bn_t& r, bn_t& s, const std::uint8_t* sig_p, std::size_t len) {
+    bool bRes{false};
+
+    auto* signature = d2i_ECDSA_SIG(nullptr, &sig_p, static_cast<long>(len));
+    if (signature == nullptr) {
+        log_error("d2i_ECDSA_SIG");
+    } else {
+        const auto* rbn = ECDSA_SIG_get0_r(signature);
+        const auto* sbn = ECDSA_SIG_get0_s(signature);
+
+        bRes = BN_bn2binpad(rbn, r.data(), static_cast<int>(r.size())) != -1;
+        bRes = bRes && BN_bn2binpad(sbn, s.data(), static_cast<int>(s.size())) != -1;
+        if (!bRes) {
+            log_error("BN_bn2binpad");
+        }
+    }
+
+    ECDSA_SIG_free(signature);
+    return bRes;
+};
+
+std::vector<Certificate_ptr> load_certificates(const char* filename) {
+    assert(filename != nullptr);
+
+    std::vector<Certificate_ptr> result{};
+    auto* store = OSSL_STORE_open(filename, UI_null(), nullptr, nullptr, nullptr);
+
+    if (store != nullptr) {
+        while (OSSL_STORE_eof(store) != 1) {
+            auto* info = OSSL_STORE_load(store);
+
+            if (info != nullptr) {
+                if (OSSL_STORE_error(store) == 1) {
+                    log_error("OSSL_STORE_load");
+                } else {
+                    const auto type = OSSL_STORE_INFO_get_type(info);
+
+                    if (type == OSSL_STORE_INFO_CERT) {
+                        // get a copy of the certificate
+                        auto cert = OSSL_STORE_INFO_get1_CERT(info);
+                        result.push_back({cert, &X509_free});
+                    }
+                }
+            }
+
+            OSSL_STORE_INFO_free(info);
+        }
+    }
+
+    OSSL_STORE_close(store);
+    return result;
+}
+
+std::string certificate_to_pem(const x509_st* cert) {
+    assert(cert != nullptr);
+
+    auto* mem = BIO_new(BIO_s_mem());
+    std::string result;
+    if (PEM_write_bio_X509(mem, cert) != 1) {
+        log_error("PEM_write_bio_X509");
+    } else {
+        BIO_flush(mem);
+
+        char* ptr{nullptr};
+        const auto size = BIO_get_mem_data(mem, &ptr);
+
+        result = std::string(ptr, size);
+    }
+    BIO_free(mem);
+    return result;
+}
+
+Certificate_ptr der_to_certificate(const std::uint8_t* der, std::size_t len) {
+    Certificate_ptr result{nullptr, nullptr};
+    const auto* ptr = der;
+    auto* cert = d2i_X509(nullptr, &ptr, static_cast<std::int64_t>(len));
+    if (cert == nullptr) {
+        log_error("d2i_X509");
+    } else {
+        result = Certificate_ptr{cert, &X509_free};
+    }
+    return result;
+}
+
+verify_result_t verify_certificate(const x509_st* cert, const CertificateList& trust_anchors,
```

Comment:
My last point, I've seen that some utilities from the libevse-security were used for certificate handling. Can't code be unified and to use the verify certificate routines from that library instead of doing this again manually? Maybe it will make code easier to maintain in the future.